Assets/Scripts/DialogueManager.cs

```diff
@@ -0,0 +1,292 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using TMPro;
+using UnityEngine;
+
+[System.Serializable]
+public class SpeakerContainer
+{
+    public string id;
+    public GameObject speakerObject;
+    public string npcVoicePath;
+
+    public SpeakerContainer(string _id, GameObject _speakerObject, string _npcVoicePath)
+    {
+        id = _id;
+        speakerObject = _speakerObject;
+        npcVoicePath = _npcVoicePath;
+    }
+}
+
+public class DialogueManager : Singleton<DialogueManager>
+{
+    public float charTypeSpeed = 0.05f;
+    public float waitLineTime = Constants.WAIT_TIME_CONVO_LINE;
+    private int activeLineIndex = 0;
+    private int startFrame;
+
+    private bool isConversing;
+    private bool autoPlaying = false;
+    private bool isTyping = false;
+    private bool skipRequest = false;
+    private bool waitingForNext = false;
+
+    public TextMeshProUGUI textMesh;
+    private Conversation currentConvo;
+    private SpeakerUI speakerUI;
+    private SpeakerContainer currentSpeaker;
+
+    private Coroutine coroutine;
+    private Coroutine currentTypingCoroutine;
+
+    public List<SpeakerContainer> allSpeakers;
+    private Dictionary<string, SpeakerContainer> speakers;
+
+    public delegate void FinishTypingEvent(string typedText);
+    public event FinishTypingEvent OnFinishTyping;
+
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        speakers = new Dictionary<string, SpeakerContainer>();
+
+        //Load all speakers into dictionary
+        foreach(SpeakerContainer speaker in allSpeakers)
+        {
+            speakers.Add(speaker.id, speaker);
+        }
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        HandleInput();
+    }
+
+    private void HandleInput()
+    {
+        if (UIManager.Instance.IsGamePaused()) return;
```

Comment:
Make sure this still works:
* enter in range of an npc
* pause
* unpause using keyboard key Z
EXPECTED BEHAVIOUR: should not start conversation

had this bug before so we should make sure this still works properly

Reply:
Ok I just tested it out and it unpaused the screen without starting the convo

Reply:
But I guess an issue now is making sure that the pause appears above the dialogue